src/provider/TwapV3.sol

```diff
@@ -0,0 +1,124 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.22;
+
+import {IUniswapV3Pool} from "@uniswap/v3-core/contracts/interfaces/IUniswapV3Pool.sol";
+import {TickMath} from "@uniswap/v4-core/src/libraries/TickMath.sol";
+import {FullMath} from "@uniswap/v4-core/src/libraries/FullMath.sol";
+import {FixedPoint96} from "@uniswap/v4-core/src/libraries/FixedPoint96.sol";
+import {OracleSource} from "./OracleSource.sol";
+
+import {AccessControl} from "@openzeppelin/contracts/access/AccessControl.sol";
+
+/**
+ * @title TwapV3
+ * @custom:version 1.0.0
+ * @author Sean Matt English (@smatthewenglish)
+ * @custom:coauthor Lorran Sutter (@LorranSutter)
+ * @custom:coauthor Dylan Moreland (@elffjs)
+ * @custom:coauthor Yevgeny Khessin (@zer0stars)
+ * @custom:coauthor Rob Solomon (@robmsolomon)
+ * @custom:contributor Allyson English (@aesdfghjkl666)
+ */
+contract TwapV3 is OracleSource, AccessControl {
+    address public immutable poolWmaticUsdc;
+    address public immutable poolWmaticDimo;
+
+    uint32 _twapIntervalUsdc;
+    uint32 _twapIntervalDimo;
+
+    uint256 constant SCALING_FACTOR = 1 ether;
+
+    constructor() {
+        _twapIntervalUsdc = 30 minutes;
+        _twapIntervalDimo = 30 minutes;
+
+        _grantRole(DEFAULT_ADMIN_ROLE, msg.sender);
+
+        poolWmaticUsdc = 0xB6e57ed85c4c9dbfEF2a68711e9d6f36c56e0FcB;
+        poolWmaticDimo = 0x41e64a5Bc929fA8E6a9C8d7E3b81A13b21Ff3045;
```

Comment:
do we want these to be passed in rather than hardcoded?

Reply:
the oracle sources are modular, so we can just swap it out if we need. the address of those pools won't change for uniswap v3, for v4 it will but thats gonna necesiate a new oracle source impl